tracing/tracing.go

```diff
@@ -0,0 +1,187 @@
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/common/version"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp"
+	"go.opentelemetry.io/otel/propagation"
+	"go.opentelemetry.io/otel/sdk/resource"
+	tracesdk "go.opentelemetry.io/otel/sdk/trace"
+	semconv "go.opentelemetry.io/otel/semconv/v1.7.0"
+	"go.opentelemetry.io/otel/trace"
+
+	"github.com/prometheus/prometheus/config"
+)
+
+const serviceName = "prometheus"
+
+// Manager is capable of building, (re)installing and shutting down
+// the tracer provider.
+type Manager struct {
+	logger       log.Logger
+	done         chan struct{}
+	config       config.TracingConfig
+	shutdownFunc func() error
+}
+
+// NewManager creates a new tracing manager.
+func NewManager(logger log.Logger) *Manager {
+	return &Manager{
+		logger: logger,
+		done:   make(chan struct{}),
+	}
+}
+
+// Run starts the tracing manager. It registers the global text map propagator and error handler.
+// It is blocking.
+func (m *Manager) Run() {
+	otel.SetTextMapPropagator(propagation.TraceContext{})
+	otel.SetErrorHandler(otelErrHandler(func(err error) {
+		level.Error(m.logger).Log("msg", "OpenTelemetry handler returned an error", "err", err)
+	}))
+	<-m.done
+}
+
+// ApplyConfig takes care of refreshing the tracing configuration by shutting down
+// the current tracer provider (if any is registered) and installing a new one.
+func (m *Manager) ApplyConfig(cfg *config.Config) error {
+	// Update only if a config change is detected.
+	if m.config == cfg.TracingConfig {
+		return nil
+	}
+
+	if m.shutdownFunc != nil {
+		if err := m.shutdownFunc(); err != nil {
+			return errors.Wrap(err, "failed to shut down the tracer provider")
+		}
+	}
+
+	// If no endpoint is set, assume tracing should be disabled.
+	if cfg.TracingConfig.Endpoint == "" {
+		m.config = cfg.TracingConfig
+		m.shutdownFunc = nil
+		otel.SetTracerProvider(trace.NewNoopTracerProvider())
+		level.Info(m.logger).Log("msg", "Tracing provider uninstalled.")
+		return nil
+	}
+
+	tp, shutdownFunc, err := buildTracerProvider(context.Background(), cfg.TracingConfig)
+	if err != nil {
+		return errors.Wrap(err, "failed to install a new tracer provider")
+	}
+
+	m.shutdownFunc = shutdownFunc
+	m.config = cfg.TracingConfig
+	otel.SetTracerProvider(tp)
+
+	level.Info(m.logger).Log("msg", "Successfully installed a new tracer provider.")
+	return nil
+}
+
+// Stop gracefully shuts down the tracer provider and stops the tracing manager.
+func (m *Manager) Stop() {
+	defer close(m.done)
+
+	if m.shutdownFunc == nil {
+		return
+	}
+
+	if err := m.shutdownFunc(); err != nil {
+		level.Error(m.logger).Log("msg", "failed to shut down the tracer provider", "err", err)
+	}
+
+	level.Info(m.logger).Log("msg", "Tracing manager stopped")
+}
```

Comment:
It doesn't really matter because there aren't any background processes anyway, but it is technically inaccurate to declare that the manager has been shut down before the channel has been closed.

```suggestion
func (m *Manager) Stop() {
	defer func() {
		close(m.done)
		level.Info(m.logger).Log("msg", "Tracing manager stopped")
	}()

	if m.shutdownFunc == nil {
		return
	}

	if err := m.shutdownFunc(); err != nil {
		level.Error(m.logger).Log("msg", "failed to shut down the tracer provider", "err", err)
	}
}
```